### DISCUSSION TITLE
Sharing code with browsers: .ts extension issue

### DISCUSSION THREAD

**Discussion by ondras** [denoland/deno#9723] [General]
Hi,

for my current Deno project, I tried re-using some of the existing code initially written for a browser. This works:

```js
import stuff from "../browser/browser-module.ts";
```

However, I run into issues when the `browser-module` depends on other files:

```js
// browser-module.ts
import another from "./another.js";
```

...because Deno requires the use of .ts extension. The whole browser-side setup works either with .js (as my .ts files are compiled down to .js) or without extensions at all, because TSC seems to be smart enough. But I am unable to use .ts (required by Deno) with TSC -- it fails with 

```
An import path cannot end with a '.ts' extension.
```

Any ideas? Thanks!

**Comment by artisonian** [comment]
Are you compiling with `tsc`? If so, I'd expect it to fail, but Deno's runtime should recognize both `.ts` and `.js` modules in the same file (for example, I can import `mustache.js` from [unpkg](https://www.unpkg.com/mustache@4.0.0/mustache.mjs) without issue).

**Comment by ondras** [comment]
The client side is compiled with `tsc` (set of .ts files =>set of .js files) and -- optionally -- bundled with `rollup` for production.



> Deno's runtime should recognize both `.ts` and `.js` modules in the same file

Do I understand correctly that you suggest compiling/bundling the client side with Deno as well (via `deno bundle`)? I would probably not need any .js files at all in this case. But I would have to rely on Deno's bundling (`rollup` seems to do a better job here) and also I would lose access to some `tsconfig.json` configuration options (`target`, for instance).

All in all, I do not think that compiling/bundling the client side code with Deno is the proper way to go.

**Reply by ondras** [reply]
> The browser has an ecmascript module loader, so why bundle?

To reduce the number of HTTP requests. 

> I think he's suggesting not compiling and not bundling at all and just referencing the .ts files directly

`tsc` is unable to resolve imports that use the `.ts` extension. (Well, at least when I tried back in 2020. This issue is old and things might have changed.)

**Comment by artisonian** [comment]
I'm not suggesting using `deno bundle` for client-side code (see #3793 for an extended discussion), but rather trying to understand your build process. If you want to include `.ts` modules compatible with both Deno and browsers, your browser build pipeline would need to transform those import statements for you. Otherwise, you might be able to use import maps on the Deno side to resolve bare imports.

**Comment by ondras** [comment]
> . If you want to include `.ts` modules compatible with both Deno and browsers, your browser build pipeline would need to transform those import statements for you.

Right! Ideally, I would love `tsc` to accept .ts imports as well (they are TypeScript, right?). But I am not aware of any way to do so.



> Otherwise, you might be able to use import maps on the Deno side to resolve bare imports.

This sounds like an interesting option. Will give them a try.

**Reply by justinmchase** [reply]
There is a huge discussion on this in the typescript repository and despite every user in the world wanting them to support `.ts` imports they will not do it because it violates a principle that they are sticking to for better or worse. No matter how many times you read the justification it won't make sense but the short version is no there is no supported way in the tyepscript compiler itself to support bare `.ts` imports.

**Comment by ondras** [comment]
> Deno's runtime should recognize both `.ts` and `.js` modules in the same file

It does, but does not really work with transpiled files; they have type information stripped from them. Here are two possible approaches, both failing at various places.

1) using .ts in import specifiers: 

```js
// deno-server.ts:
import { SomeType } from "./browser/some-type.ts"; // server side: works when compiling with Deno

// browser/some-type.ts:
import other from "./other.ts"; // client side: fails, tsc does not expect .ts
```


2) using .js in client-side import specifiers:
```js
// deno-server.ts:
import { SomeType } from "./browser/some-type.js"; // server side: fails, this is a transpiled file without type information

// browser/some-type.ts:
import other from "./other.js"; // client side: works
```

**Comment by ondras** [comment]
To sum things up: with Deno, I cannot import a tsc-compiled dependency that has other dependencies.

**Comment by kitsonk** [comment]
While this evaded me before now, I think we will be able to fix it in #4538 for `Deno.compile()`.  Would a fix there work for your @ondras?

**Comment by ondras** [comment]
> Would a fix there work for your @ondras?

No, I am not doing `Deno.compile()`, I am trying to `deno run` some code that has transitive client-side dependencies. This is a sample file: https://github.com/ondras/rri/blob/master/server/player.ts

These direct dependencies go to `../src/*.ts` which is OK, but their additional dependencies ([example](https://github.com/ondras/rri/blob/master/src/board.ts)) are declared with `.js` extensions, because that is what TSC needs. So I am currently [generating a large importmap](https://github.com/ondras/rri/blob/master/server/Makefile#L8) that converts *.js to *.ts, so that Deno can process these `import "x.js"` statements as well. Not really optimal, but probably the only way to go.

All in all, I do not think this is a bug in Deno, because I do not see a proper Deno-based solution for this scenario. All my source files are TypeScript, so they all shall have a .ts extension. Unfortunately, the need for client-side TS compilation is forcing me to use .js, Deno-incomaptible.

**Comment by lucacasonato** [comment]
Is anything here actionable for us? If not I would like to close and move the discussion to https://github.com/denoland/deno/discussions or https://discord.gg/deno.

**Comment by jsejcksn** [comment]
> Is anything here actionable for us? If not I would like to close and move the discussion to https://github.com/denoland/deno/discussions or https://discord.gg/deno.

I'm ok with that. Please post here the link to the new discussion if you move it.

**Comment by Higgs1** [comment]
Hello,

I/we would very much like to write Web-compatible modules in TypeScript (that only utilize Web compatible APIs) to JavaScript so that it works in Web browsers and/or other runtimes that can run ES2015 compatible JavaScript. Has there been any progress on this? I/we are fine with _any_ build pipeline that does this, as long as it doesn't involve sentient beings changing ".ts" to ".js" (this is what computer software is for).

**Reply by jsejcksn** [reply]
@Higgs1 See https://github.com/denoland/dnt

**Reply by Higgs1** [reply]
The title of the project is "Deno to Node Transform" and the description "Deno to npm package build tool" but reading through it a little more it does appear to emit ESM - and it uses Deno and not Node! Thank you, I'll have to look into getting that project to work for us since we weren't yet able to figure out how to accomplish this with the provided documentation (converting web-compatible TS Modules -> ESM, without anything to do with Node/NPM). I'll report back here with my findings. Any additional assistance would also be much appreciated.

**Reply by jtoppine** [reply]
> Has there been any progress on this? I/we are fine with _any_ build pipeline that does this, as long as it doesn't involve sentient beings changing ".ts" to ".js" (this is what computer software is for).

esbuild should do exactly what you want.  https://deno.land/x/esbuild Works well with JS, never tried with TS personally but apparently it should support TS just fine.

**Reply by Higgs1** [reply]
@jtoppine Thanks for sharing. I've read through the documentation on https://esbuild.github.io/ but I still can't seem to locate any option that enables conversion between working Deno TS -> ESM. It appears as if we'd have to write TypeScript with .js imports (that don't exist on the filesystem, only in imaginary TSC land) and then run multiple pipelines:

* esbuild produce .d.ts (or maybe JSDOC??) files so that Deno can see type information
* esbuild to convert source to JavaScript modules (at the .js locations)
* esbuild others as desired

While that does sound like it would work, it's not super ideal as right now we're able to code in working Deno TypeScript (*with* all the good typings and such) and are able to just test/use it with deno-test, deno-bench, deno-run, etc - _without any intermediate steps_. From what I can gather about esbuild, we'd have to compile it to Deno TS first and have to set up watch scripts in our build environment before we can feed it to Deno and also additionally have more files moving around. I'm not certain how that is much of an improvement over our current situation - `sed` scripts - other than the fact that its not sed.

At the very least it seems it can put us in the same position as `tsc` does, with the massive benefits that it doesn't require Node or NPM, has fewer dependencies, and already comes with a plugin system (that we could potentially add a .ts -> .js transformer into).

Still, wouldn't it be simpler just to add a command line flag to esbuild (or forbiddenly, tsc)? Multiple users have already submitted PRs to do just that but have been denied (repeatedly). Iff we do make a plugin for esbuild, where might be the best place to upload it so that it could have the most impact? (other alternative solutions are still appreciated, thank you!!!)

**Reply by ayame113** [reply]
Out of pure curiosity, why not use [deno_emit](https://github.com/denoland/deno_emit)?
This library converts TypeScript files written for Deno into JavaScript.

```ts
import { bundle } from "https://deno.land/x/emit/mod.ts";
const { code } = await bundle(new URL(import.meta.resolve("./path/to/entrypoint.ts")));
// console.log(code);
await Deno.writeTextFile("./output.js", code);
```

**Reply by Higgs1** [reply]
Thanks for the link @ayame113 . I've just now read up on [deno_emit](https://github.com/denoland/deno_emit)'s `emit` function, which does seem to transpile TypeScript modules to JavaScript modules but similarly to `esbuild`, I can't seem to get it to convert working Deno TypeScript modules to working JavaScript modules. While `deno_emit` does seem to be even lighter than esbuild (which is a big plus!), I don't think it provides a plugins functionality, so we won't be able to write a plugin to do the conversion from .ts to .js without a `sed` script or similarly language-dumb solution. Unless we've missed something of course!

**Reply by ayame113** [reply]
What you want to do is run the ES module in the browser without bundling the module written for Deno? (Sorry if I didn't understand your intention, I read English with Google Translate.)

I have tried to do this in the past and have come to the conclusion that it is difficult. Since `mod.ts` and `mod.js` can coexist in Deno, simply converting `mod.ts` and outputting to `mod.js` will result in a conflicting output file.

I gave up on this method and decided to use another method. in short,
1. Compile TypeScript files with extension .ts to create JavaScript files with extension **.ts** (In other words, it does not change the file name at all.)
2. Distribute the JavaScript file created in 1 from the server with `Content-Type: application/javascript`
3. `<script src="./mod.ts" type="module">` and `import "mod.ts"` work on the browser

The point of this method is that there is no need to rewrite the import statement path at all.
I wrote a library called [ts-serve](https://github.com/ayame113/ts-serve) that can complete these processes on the server.

If you want to create a build pipeline that does something similar to this, it feels like a very viable thing to do.

**Reply by Higgs1** [reply]
That's actualy a pretty interesting and novel solution! And thank you for spending all this time trying to assist a non-Japanese speaker. But yes, that's the intended goal. 

Rant begin.

But I can't give up. TypeScript and JavaScript both have modules now. Using the same physical syntax. The runtimes for Deno, Bun, GJS, and web browsers all use the same core API that can do among other things: math, string ops, crypto, compression, console, http requests, streams, async/await, workers, GPU. Deno can seamlessly use TypeScript and JavaScript at the same time, and even load Node modules. Bun can run Node modules. TSC and Deno and others can transpile TypeScript to JavaScript. ESBuild, deno-emit, deno-bundle, webpack, et.al. can all bundle a whole tree of mixed TypeScript and JavaScript into _n_ files which work in the web browser. Deno, Bun, TSC all work on the module level. CoffeeScript transpiles import statements .coffee -> .js

    CoffeeScript transpiles import statements .coffee -> .js

There is just one tiny missing puzzle piece, and it is a _VERY_ highly requested feature, but it has to be attached to a transpiled because only a transpiler has all the required knowledge. Just imagine - coding in TypeScript once, and then deploying - using a _static_ server to either browser or Node or Deno or anyother runtime that supports loading external modules. TypeScript seems to be so popular (I mean, it's a built-in function for Deno), yet why are developers forced to run their codebase through the complex maze that is webpack or rollup? Why then do browsers even have module import functionality, or to be able to load multiple javascript files when we should "just bundle it" so it can "load faster". Why must functionality be rewritten a second time for consumers who don't want to use those toolings? Imagine transpiling to working JavaScript, and _then_ tree shaking! Tools would (probably) be so much easier to write and maintain. 

Also it's a lot of work trying to teach somebody about the Node/NPM/webpack ecosystem. How many greenhorns have looked at a cool feature on a website, say... hoving over an image to make it appear wobbly and glitchy and wondered how they could achieve that look? The answer is simple: set up a container, download node, npm,  maybe gulp or grunt, run all the updates, probably get scss too, write their website in not-quite-JavaScript, not-quite-CSS, not-quite-HTML, set up a dev environment that watches your files for changes and automatically compiles the code, debug, set up the transpilers, pack it, minify it, then serve it on produciton. All of the old web dev books that tell people that they can just open up a text editor, write some javascript, refresh the page pointing to their nginx server, and debug using view source or FireBug... just doesn't apply anymore.

At this rate I think we've made a mistake starting a project(s) with TypeScript. It just doesn't work with things! JavaScript seems to be the way to go, always has been. Use JSDoc for typings!

**Reply by jsejcksn** [reply]
@Higgs1 I infer that you also don't necessarily want to bundle, leaving that concern to a downstream handler. Here's an example based on what @ayame113 [commented](https://github.com/denoland/deno/discussions/9723#discussioncomment-4898200):

```ts
import { a } from "./mod.ts";
import { b } from "./mod.js";
import { c } from "./mod.ts.js";

console.log({ a, b, c });
```

In that case, what should be the output in accordance with your imagined specifier rewriting system?

Further (Socratic): Have you thought through conflict cases and considered a ruleset which can handle all such cases?

**Reply by Higgs1** [reply]
Hi @jsejcksn ,

With the input from several others, I'm pretty confident that I've thought through a wide variety of cases: I don't want any particular specifier rewriting system - not yet anyway, others may standardize a convention later if they want. As you've mentioned elsewhere, the code that's written should follow the conventions of the compiler. Right now, the only puzzle piece I'm interested in is a hypothetical transpiler that emits valid, working JavaScript that conforms to WHATWG's module loading system in a deterministic way. Here are a few possible rewriting systems that I've come up with:

1. Output to a different path, don't rewrite any specifiers.
```javascript
import { a } from "./mod.ts";
import { b } from "./mod.js";
import { c } from "./mod.ts.js";

console.log({ a, b, c });
````
2. Hash all specifiers. Note: Produces collisions.
```javascript
import { a } from "./e6eb0057c923c4a224c4d05275cf35cf";
import { b } from "./74c68ff4eff48c7f153caba19b3307cb";
import { c } from "./844aa07b6e487967e3f135148baa9126";

console.log({ a, b, c });
```
3. Just append .js to everything. Guarenteed to be unique. Causes issues on systems where path length is limited. 
```javascript
import { a } from "./mod.ts.js";
import { b } from "./mod.js.js";
import { c } from "./mod.ts.js.js";

console.log({ a, b, c });
````

Rewriting external imports seems to be a completely tangental yet equally possible problem. One can probably utilize annotations on the remote side (like `@ sourceMappingURL` - "hey, my transpiled ESM code is at..."), an HTTP header if served over HTTP (like X-TypeScript-Types), annotations when importing (like `@deno-types` - "the ESM code for this TS resource is at..."), specifying them in a sort of input mapping file, specifiying a specific pre-defined ruleset (by command argument for example), or if all else fails just pull them and compile them locally alongside the original codebase; or a combination of all the above.

**Reply by jsejcksn** [reply]
@Higgs1 Based on your interest in this, I also want to make sure you're aware of a proposal which relates to some of the concerns you've shared: [ECMAScript proposal: Type Annotations](https://github.com/tc39/proposal-type-annotations)

**Reply by Higgs1** [reply]
Oh thank you @jsejcksn , I was not aware of that proposal but that's something I've been looking for since 2012. Like JSDoc, but defined as part of the syntax. We can finally get rid of TypeScript. Unfortunately I have absolutely no idea how to contribute to that. I had started a discussion about types in JS on webplatform.org earlier (/ a decade ago) but it didn't go anywhere... I've increased the project's star count at least.

Additionally, a note on our progress: we're working on _coercing_ `esbuild` to do what we want, using a separate output directory (method no. 1 above). There's a lot of steps that aren't obvious because we want a .d.ts file, and a minified file, and we need to process the intermediate source mappings. 

Apparently Microsoft decrees that JavaScript files use the .ts extension now (and thus far we're obeying!). It's still pretty madenning that the JS community is so close, yet so far way...

**Reply by Higgs1** [reply]
Update: any setup that involves leaving the files as .ts extension makes practically every service serve the files as either `text/plain` or `video/mp2t` which is incorrect. Changing the content type is not possible on any of these hosting services.

We're at a complete loss. What can be done? Should I post something on the Type Annotations proposal issue tracker? Why is the community so intent on starting a new browser war?

**Comment by ondras** [comment]
Hi,

FYI: tsc now [accepts .ts imports](https://github.com/microsoft/TypeScript/pull/51669), so this issue/discussion is no longer relevant - at least in my original case.

**Reply by nayeemrmn** [reply]
Out of curiosity, what tool are you using to build that TS for the browser?

**Reply by ondras** [reply]
esbuild seems to bundle that. Not sure what minimum version would work - 0.17.18 is the one I successfully tried.

**Reply by ondras** [reply]
(my main problem with .ts imports was VS Code, though - which now works with the `"allowImportingTsExtensions": true` option)
